src/changeset.rs

```diff
@@ -0,0 +1,184 @@
+use std::iter::FromIterator;
+use std::ops::AddAssign;
+
+use join::Join;
+use storage::{DenseVecStorage, UnprotectedStorage};
+use world::{Entity, Index};
+
+/// Change set that can be collected from an iterator, and joined on for easy application to
+/// components.
+///
+/// ### Example
+///
+/// ```rust
+/// # extern crate specs;
+/// # use specs::prelude::*;
+///
+/// pub struct Health(i32);
+///
+/// impl Component for Health {
+///     type Storage = DenseVecStorage<Self>;
+/// }
+///
+/// # fn main() {
+/// # let mut world = World::new();
+/// # world.register::<Health>();
+///
+/// let a = world.create_entity().with(Health(100)).build();
+/// let b = world.create_entity().with(Health(200)).build();
+///
+/// let changeset = [(a, 32), (b, 12), (b, 13)]
+///     .iter()
+///     .cloned()
+///     .collect::<ChangeSet<i32>>();
+/// for (health, modifier) in (&mut world.write::<Health>(), &changeset).join() {
+///     health.0 -= modifier;
+/// }
+/// # }
+/// ```
+pub struct ChangeSet<T> {
+    mask: BitSet,
+    inner: DenseVecStorage<T>,
+}
+
+impl<T> ChangeSet<T> {
+    /// Create a new change set
+    pub fn new() -> Self {
+        ChangeSet {
+            mask: BitSet::default(),
+            inner: DenseVecStorage::default(),
+        }
+    }
+
+    /// Add a value to the change set. If the entity already have a value in the change set, the
+    /// incoming value will be added to that.
+    pub fn add(&mut self, entity: Entity, value: T)
+    where
+        T: AddAssign,
+    {
+        if self.mask.contains(entity.id()) {
+            unsafe {
+                *self.inner.get_mut(entity.id()) += value;
+            }
+        } else {
+            unsafe {
+                self.inner.insert(entity.id(), value);
+            }
+            self.mask.add(entity.id());
+        }
+    }
+
+    /// Clear the changeset
+    pub fn clear(&mut self) {
+        for id in &self.mask {
+            unsafe {
+                self.inner.remove(id);
+            }
+        }
+        self.mask.clear();
+    }
+}
+
+impl<T> FromIterator<(Entity, T)> for ChangeSet<T>
+where
+    T: AddAssign,
+{
+    fn from_iter<I: IntoIterator<Item = (Entity, T)>>(iter: I) -> Self {
+        let mut changeset = ChangeSet::new();
+        for (entity, d) in iter {
+            changeset.add(entity, d);
+        }
+        changeset
+    }
+}
+
+impl<T> Extend<(Entity, T)> for ChangeSet<T>
+where
+    T: AddAssign,
+{
+    fn extend<I: IntoIterator<Item = (Entity, T)>>(&mut self, iter: I) {
+        for (entity, d) in iter {
+            self.add(entity, d);
+        }
+    }
+}
+
+impl<'a, T> Join for &'a mut ChangeSet<T> {
+    type Type = &'a mut T;
+    type Value = &'a mut DenseVecStorage<T>;
+    type Mask = &'a BitSet;
+
+    fn open(self) -> (Self::Mask, Self::Value) {
+        (&self.mask, &mut self.inner)
+    }
+
+    unsafe fn get(v: &mut Self::Value, id: Index) -> Self::Type {
+        let value: *mut Self::Value = v as *mut Self::Value;
```

Comment:
I don't think this is necessary.

Reply:
It is taken from the Join on &mut Storage, it's required for the compiler to understand that v is mutable.

Reply:
It fails to validate the lifetime without it.

Reply:
The basic issue is that it fails to match &mut on the function signature to &'a for the Join implementation, and it's not possible to set &'a on the function signature because it would not match the trait then.

Reply:
Ah, right.